guava/src/com/google/common/collect/ImmutableSet.java

```diff
@@ -698,9 +699,19 @@ static int maxRunBeforeFallback(int tableSize) {
   }
 
   /**
-   * Default implementation of the guts of ImmutableSet.Builder, creating an open-addressed hash
-   * table and deduplicating elements as they come, so it only allocates O(max(distinct,
-   * expectedCapacity)) rather than O(calls to add).
+   * Default implementation of the guts of ImmutableSet.Builder, creating many open-addressed hash
+   * tables and deduplicating elements within each table as they come.
+   *
+   * <p>We do not deduplicate all elements as they come as doing so requires a lot of hash table
+   * resizing if many elements are being added. Instead, we will deduplicate and hash all elements
+   * until the table needs resizing. Then we will create a new table without copying the previous
+   * table, and hash and deduplicate only newly added elements until the new table is full, and then
+   * merging the deduplicated tables together at the end. This way, adding n elements, m of which
+   * are distinct, saves about O(m/2) rehashes of elements compared to deduplicating all elements
+   * and resizing a single hashtable.
+   *
+   * <p>while worst case, we allocate O(n) during building, given uniformly distributed duplicates
```

Comment:
I believe you're underselling this, actually: I think the allocation is still O(n) worst case, not O(m).

AFAICT, the worst-case input for this implementation is e.g.

1, 1, 2, 1, 2, 3, 4, 1, 2, 3, 4, 5, 6, 7, 8, ....

in which case overflowTables contains

{1}, {1, 2}, {1, 2, 3, 4}, {1, 2, 3, 4, 5, 6, 7, 8}

which has "maximum duplication": every element that can be duplicated is duplicated as many times as possible.  But we've still only allocated O(m): m = n/2.

Reply:
m is the number of distinct elements, n is the number of puts.

The above cause you mentioned would still end up with O(n) total memory in the worst case, since we keep every one of those {1}, {1,2} .... tables, making total of O(n) space.
It's true that the largest/latest overflow table, and the final allocation for the set would only be O(m) = O(n/2) though.